src/Servers/HttpSys/src/RequestProcessing/Response.cs

```diff
@@ -527,111 +527,89 @@ private static bool Matches(string knownValue, StringValues input)
-                        }
-
-                        for (int headerValueIndex = 0; headerValueIndex < headerValues.Count; headerValueIndex++)
-                        {
-                            // Add Name
-                            bytes = HeaderEncoding.GetBytes(headerName);
-                            unknownHeaders[_nativeResponse.Response_V1.Headers.UnknownHeaderCount].NameLength = (ushort)bytes.Length;
-                            gcHandle = GCHandle.Alloc(bytes, GCHandleType.Pinned);
-                            pinnedHeaders.Add(gcHandle);
-                            unknownHeaders[_nativeResponse.Response_V1.Headers.UnknownHeaderCount].pName = (byte*)gcHandle.AddrOfPinnedObject();
-
-                            // Add Value
-                            headerValue = headerValues[headerValueIndex] ?? string.Empty;
-                            bytes = HeaderEncoding.GetBytes(headerValue);
-                            unknownHeaders[_nativeResponse.Response_V1.Headers.UnknownHeaderCount].RawValueLength = (ushort)bytes.Length;
-                            gcHandle = GCHandle.Alloc(bytes, GCHandleType.Pinned);
-                            pinnedHeaders.Add(gcHandle);
-                            unknownHeaders[_nativeResponse.Response_V1.Headers.UnknownHeaderCount].pRawValue = (byte*)gcHandle.AddrOfPinnedObject();
-                            _nativeResponse.Response_V1.Headers.UnknownHeaderCount++;
-                        }
+                        // Add Name
+                        bytes = HeaderEncoding.GetBytes(headerName, writer);
+                        unknownHeaders[_nativeResponse.Response_V1.Headers.UnknownHeaderCount].NameLength = (ushort)bytes.Length;
+                        unknownHeaders[_nativeResponse.Response_V1.Headers.UnknownHeaderCount].pName = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(bytes));
+
+                        // Add Value
+                        headerValue = headerValues[headerValueIndex] ?? string.Empty;
+                        bytes = HeaderEncoding.GetBytes(headerValue, writer);
+                        unknownHeaders[_nativeResponse.Response_V1.Headers.UnknownHeaderCount].RawValueLength = (ushort)bytes.Length;
+                        unknownHeaders[_nativeResponse.Response_V1.Headers.UnknownHeaderCount].pRawValue = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(bytes));
+                        _nativeResponse.Response_V1.Headers.UnknownHeaderCount++;
                     }
-                    else if (headerPair.Value.Count == 1)
+                }
+                else if (headerPair.Value.Count == 1)
+                {
+                    headerValue = headerValues[0] ?? string.Empty;
+                    bytes = HeaderEncoding.GetBytes(headerValue, writer);
+                    pKnownHeaders[lookup].RawValueLength = (ushort)bytes.Length;
+                    pKnownHeaders[lookup].pRawValue = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(bytes));
```

Comment:
Is this the cleanest way to do this?

Reply:
As in get to the pointer within a `Span<T>`? Yes. This is a very advanced scenario, mostly used in interop, so making it cleaner isn't worth a lot of effort.